plugins/openai/src/llm.ts

```diff
@@ -0,0 +1,659 @@
+            role: llm.ChatRole.ASSISTANT,
+            toolCalls: this.fncCtx,
+          },
+          index: choice.index,
+        },
+      ],
+    };
+  }
+}
+
+const buildMessage = (msg: llm.ChatMessage, cacheKey: any) => {
+  const oaiMsg: Partial<OpenAI.ChatCompletionMessageParam> = {};
+
+  switch (msg.role) {
+    case llm.ChatRole.SYSTEM:
+      oaiMsg.role = 'system';
+      break;
+    case llm.ChatRole.USER:
+      oaiMsg.role = 'user';
+      break;
+    case llm.ChatRole.ASSISTANT:
+      oaiMsg.role = 'assistant';
+      break;
+    case llm.ChatRole.TOOL:
+      oaiMsg.role = 'tool';
+      if (oaiMsg.role === 'tool') {
+        oaiMsg.tool_call_id = msg.toolCallId;
+      }
+      break;
+  }
+
+  if (typeof msg.content === 'string') {
+    oaiMsg.content = msg.content;
+  } else if (
+    ((c?: llm.ChatContent | llm.ChatContent[]): c is llm.ChatContent[] => {
+      return (c as llm.ChatContent[]).length !== undefined;
+    })(msg.content)
+  ) {
+    oaiMsg.content = msg.content.map((c) => {
+      if (typeof c === 'string') {
+        return { type: 'text', text: c };
+      } else if (
+        // typescript type guard for determining ChatAudio vs ChatImage
+        ((c: llm.ChatAudio | llm.ChatImage): c is llm.ChatImage => {
+          return (c as llm.ChatImage).image !== undefined;
+        })(c)
+      ) {
+        return buildImageContent(c, cacheKey);
+      } else {
+        throw new Error('ChatAudio is not supported');
+      }
+    }) as OpenAI.ChatCompletionContentPart[];
+  }
+
+  // make sure to provide when function has been called inside the context
+  // (+ raw_arguments)
+  if (msg.toolCalls && oaiMsg.role === 'assistant') {
+    oaiMsg.tool_calls = Object.entries(msg.toolCalls).map(([name, func]) => ({
+      id: msg.toolCallId!,
+      type: 'function' as const,
+      function: {
+        name: name,
+        arguments: func.rawParams,
+      },
+    }));
+  }
+
+  return oaiMsg as OpenAI.ChatCompletionMessageParam;
+};
+
+const buildImageContent = (image: llm.ChatImage, cacheKey: any) => {
+  if (typeof image.image === 'string') {
+    // image url
+    return {
+      type: 'image_url',
+      image_url: {
+        url: image.image,
+        detail: 'auto',
+      },
+    };
+  } else {
+    if (!image.cache[cacheKey]) {
+      // inside our internal implementation, we allow to put extra metadata to
+      // each ChatImage (avoid to reencode each time we do a chatcompletion request)
+      // TODO(nbsp): resize image and encode to base64
+    }
+
+    return {
+      type: 'image_url',
+      image_url: {
+        url: `data:image/jpeg;base64,${image.cache[cacheKey]}`,
+      },
+    };
+  }
+};
+
+// unique object identifier, used for cache
+// https://stackoverflow.com/a/43963612
+const id = (() => {
+  let currentId = 0;
+  const map = new WeakMap();
+
+  return (object: any) => {
+    if (!map.has(object)) {
+      map.set(object, ++currentId);
+    }
+
+    return map.get(object);
+  };
+})();
```

Comment:
No need to be exactly like Python, feel free to generate an ID inside the constructor